internal/controller/reinstall/reinstall.go

```diff
@@ -372,3 +379,279 @@ func (r *ReinstallHandler) finalizeReinstallRequest(
+		updateRequired := false
+
+		if preservationDataBackedupCondition != nil {
+			updateRequired = setReinstallStatusCondition(clusterInstance, *preservationDataBackedupCondition)
+		}
+
+		if clusterIdentityDetectedCondition != nil {
+			if changed := setReinstallStatusCondition(clusterInstance, *clusterIdentityDetectedCondition); changed {
+				updateRequired = true
+			}
+		}
+
+		if updateRequired {
+			result.Requeue = true
+			updateErr := conditions.PatchCIStatus(ctx, r.Client, clusterInstance, patch)
+			if updateErr != nil {
+				result.RequeueAfter = requeueWithShortInterval
+				err = logAndWrapUpdateFailure(log, err, updateErr,
+					"failed to update reinstall preservation condition(s)")
+			}
+		}
+	}()
+
+	if clusterInstance.Spec.Reinstall.PreservationMode == v1alpha1.PreservationModeNone {
+		preservationDataBackedupCondition = reinstallPreservationDataBackedupConditionStatus(
+			metav1.ConditionFalse, v1alpha1.PreservationNotRequired, "PreservationMode is set to None.")
+		clusterIdentityDetectedCondition = reinstallClusterIdentityDataDetectedConditionStatus(
+			metav1.ConditionFalse, v1alpha1.PreservationNotRequired, "PreservationMode is set to None.")
+		return
+	}
+
+	preservationDataBackedupCondition = findReinstallStatusCondition(clusterInstance,
+		v1alpha1.ReinstallPreservationDataBackedup)
+	clusterIdentityDetectedCondition = findReinstallStatusCondition(clusterInstance,
+		v1alpha1.ReinstallClusterIdentityDataDetected)
+
+	setInProgress := false
+	if preservationDataBackedupCondition == nil ||
+		preservationDataBackedupCondition.Reason == string(v1alpha1.Initialized) {
+		preservationDataBackedupCondition = reinstallPreservationDataBackedupConditionStatus(
+			metav1.ConditionFalse, v1alpha1.InProgress, "Data is being preserved")
+		setInProgress = true
+	}
+
+	if clusterIdentityDetectedCondition == nil ||
+		clusterIdentityDetectedCondition.Reason == string(v1alpha1.Initialized) {
+		clusterIdentityDetectedCondition = reinstallClusterIdentityDataDetectedConditionStatus(
+			metav1.ConditionFalse, v1alpha1.InProgress, "Data is being preserved")
+		setInProgress = true
+	}
+
+	if setInProgress {
+		result.Requeue = true
+		return
+	}
+
+	// Determine if resources have been preserved already
+	preservationExecuted := false
+	switch preservationDataBackedupCondition.Reason {
+	case string(v1alpha1.Failed):
+		err = errors.New(preservationDataBackedupCondition.Message)
+		return
+
+	case string(v1alpha1.Completed), string(v1alpha1.DataUnavailable):
+		preservationExecuted = true
+	}
+
+	clusterIDDataDetectionExecuted := false
+	switch clusterIdentityDetectedCondition.Reason {
+	case string(v1alpha1.Failed):
+		err = errors.New(clusterIdentityDetectedCondition.Message)
+		return
+
+	case string(v1alpha1.DataAvailable), string(v1alpha1.DataUnavailable):
+		clusterIDDataDetectionExecuted = true
+	}
+
+	if preservationExecuted && clusterIDDataDetectionExecuted {
+		return
+	}
```

Comment:
Why do any if this checking on the existing conditions?
I'm generally not a fan of using conditions to gate behavior, as a controller you should generally always be comparing desired state to current state rather than using this as a release.

Why not check and try to do the backup every time? Maybe it will be a no-op, but it should be more correct, right?

Reply:
Refactored to run preservation & restore each time.